From 83d1d4b9314aefe5f27cc24d5e4343f616357db1 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Mon, 17 Jun 2024 08:53:35 +0800
Subject: [PATCH 37/98] compositor: Fix clone mode error

Avoid multiple globals to maintain a single panel and background.

Tested on RK3588 EVB with:
Add to weston.ini:
[output]
name=HDMI-A-2

[output]
name=HDMI-A-1
clone-of=HDMI-A-2

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 frontend/main.c             |  2 +-
 libweston/backend-drm/drm.c | 16 ++++++++++++++++
 libweston/compositor.c      | 16 ++++++++++++++++
 3 files changed, 33 insertions(+), 1 deletion(-)

diff --git a/frontend/main.c b/frontend/main.c
index a4c0a58e..43ca5b1b 100644
--- a/frontend/main.c
+++ b/frontend/main.c
@@ -3485,7 +3485,7 @@ drm_try_attach(struct weston_output *output,
 	unsigned i;
 
 	/* try to attach remaining heads, this probably succeeds */
-	for (i = 1; i < add->n; i++) {
+	for (i = 0; i < add->n; i++) {
 		if (!add->heads[i])
 			continue;
 
diff --git a/libweston/backend-drm/drm.c b/libweston/backend-drm/drm.c
index f99cc7cc..5ff7696a 100644
--- a/libweston/backend-drm/drm.c
+++ b/libweston/backend-drm/drm.c
@@ -2113,6 +2113,8 @@ drm_output_detach_head(struct weston_output *output_base,
 		       struct weston_head *head_base)
 {
 	struct drm_output *output = to_drm_output(output_base);
+	struct drm_backend *b = output->backend;
+	struct drm_device *device = b->drm;
 	struct drm_head *head = to_drm_head(head_base);
 
 	if (!output_base->enabled)
@@ -2121,6 +2123,20 @@ drm_output_detach_head(struct weston_output *output_base,
 	/* Drop connectors that should no longer be driven on next repaint. */
 	wl_list_remove(&head->disable_head_link);
 	wl_list_insert(&output->disable_head, &head->disable_head_link);
+
+	if (!wl_list_length(&output_base->head_list))
+		return;
+
+	/* XXX: ensure the configuration will work.
+	 * This is actually impossible without major infrastructure
+	 * work. */
+
+	/* Need to go through modeset to add connectors. */
+	/* XXX: Ideally we'd do this per-output, not globally. */
+	/* XXX: Doing it globally, what guarantees another output's update
+	 * will not clear the flag before this output is updated?
+	 */
+	drm_device_recovery_required(device);
 }
 
 static const struct weston_drm_format_array *
diff --git a/libweston/compositor.c b/libweston/compositor.c
index e0fd974b..582a7bee 100644
--- a/libweston/compositor.c
+++ b/libweston/compositor.c
@@ -6370,6 +6370,14 @@ bind_output(struct wl_client *client,
 static void
 weston_head_add_global(struct weston_head *head)
 {
+	struct weston_head *tmp_head;
+
+	/* Avoid multiple globals to maintain a single panel and background */
+	wl_list_for_each(tmp_head, &head->output->head_list, output_link) {
+		if (tmp_head->global)
+			return;
+	}
+
 	head->global = wl_global_create(head->compositor->wl_display,
 					&wl_output_interface, 4,
 					head, bind_output);
@@ -6772,6 +6780,10 @@ weston_output_attach_head(struct weston_output *output,
 {
 	char *head_names;
 
+	/* Already attached */
+	if (head->output == output)
+		return 0;
+
 	if (!wl_list_empty(&head->output_link))
 		return -1;
 
@@ -6844,6 +6856,10 @@ weston_head_detach(struct weston_head *head)
 			free(head_names);
 
 			weston_output_emit_heads_changed(output);
+
+			head = weston_output_get_first_head(output);
+			if (head)
+				weston_head_add_global(head);
 		}
 	}
 }
-- 
2.39.5

